### PR TITLE
[MRG] Fix GUI MPI available cores

### LINKED ISSUE #870: MPI on HPC
I've been testing out the GUI on Brown's HPC system (OSCAR). Running simulations with the MPI backend is not working because it's requesting too many processors than the instance allows. The node that my instance is running on has 48 cores, but my instance is not allotted access to all the node's cores. 

https://github.com/jonescompneurolab/hnn-core/blob/18830b53c4602c8e0f9b89502222b27799c5e3ae/hnn_core/gui/gui.py#L1916-L1918
The GUI initializes the backend at the lines above using `multiprocessing.cpu_count`, which returns the node's total but not my instance's allotment. 

The joblib backend allows you to specify the number of cores in with the GUI. Is there a reason why this option is not exposed for MPI?

*joblib options*
Screenshot 2024-08-22 at 5 27 07 PM

*MPI options*
Screenshot 2024-08-22 at 5 26 40 PM

This [stack overflow answer](https://stackoverflow.com/a/55423170) also has a way to get the number of available cores instead of total.

### PR DESCRIPTION
**Pull request by gtdang**
The GUI MPI was not working on HPC systems where the number of cores partitioned to an instance is less than the total cores of a node. 

This fix uses the number of available cores to the process instead of the total number of cores of the system. 

**Work done**
1. Use psutil to get number of available cores instead of total system cores for Linux and Windows. Mac will still use total system cores.
2. Exposed the number of cores selection for the MPI backend. Default is set to 1 and the maximum is set to the number of available cores. 
3. Fixed the issue where the backend sub-options (n cores) was not displaying for the default (joblib) on startup  

Screenshot 2024-08-26 at 1 49 28 PM

closes #870

### CODE REVIEW

File: hnn_core/gui/gui.py [hunk 1]
```diff
@@ -320,6 +321,9 @@ def __init__(self, theme_color="#802989",
         # load default parameters
         self.params = self.load_parameters(network_configuration)
 
+        # Number of available cores
+        self.n_cores = self._available_cores()
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:325]: Usually, I've been told that one shouldn't use all the cores available when doing multiprocessing ... the user should be told how many cores are available and how many of those they want to use.

Comment (gtdang) [reply hnn_core/gui/gui.py:325]: Yes agreed. I documented the GUI's behavior in issue https://github.com/jonescompneurolab/hnn-core/issues/870

Using the joblib backend the number of cores is exposed and adjustable. With the MPI backend it is not. MPI uses the total number of cores-1. I was curious if there was a reason it was designed this way.

**Comment by gtdang** [comment]
Ok. I've exposed the number of cores selection for MPI. The default is set to 1 and the maximum should be the total number of cores available to the process rather than the total number of cores on the system. 

Note that core availability applies to Linux and Windows systems, Mac will just show the total system cores. This shouldn't be an issue given that this issue mostly applies to running on HPC nodes. Also the user can now adjust the number of cores for MPI manually.

**Comment by rythorpe** [comment]
Sorry, I didn't notice this until now. I just made a [comment](https://github.com/jonescompneurolab/hnn-core/issues/870#issuecomment-2311030156) on the issue page. I'm not opposed to this change, but also don't think it's absolutely necessary to expose the parallel backend API in the GUI with the sole reason that it adds yet another variable for new users to trip over.

**Comment by dylansdaniels** [comment]
@rythorpe are you saying remove the Backend option from the GUI entirely? We do switch the backend from joblib to MPI via the GUI in the workshops to speed things up for users and make them aware of that functionality, so I think we do need it. Though we can simplify it. I also weighed in on the [issue](https://github.com/jonescompneurolab/hnn-core/issues/870#issuecomment-2311149512) page :)

**Comment by jasmainak** [comment]
I guess it's fair to say we just use MPIBackend and default to joblib if not available ... so no need to expose n_cores for joblibbackend and keep it simple

**Comment by gtdang** [comment]
I've been testing the workshop instances on Oscar (8 cores). I'm finding that MPI will fail on runs with 6-8 cores specified with the message below: 
```
--------------------------------------------------------------------------
A request was made to bind to that would result in binding more
processes than cpus on a resource:

   Bind to:     CORE
   Node:        node1717
   #processes:  2
   #cpus:       1

You can override this protection by adding the "overload-allowed"
option to your binding directive.
--------------------------------------------------------------------------
```

In instance is partitioned 8 cores on a 32 core node.  
```
(hnn_env) [gdang2@node1717 ~]$ nproc
8
(hnn_env) [gdang2@node1717 ~]$ python
Python 3.11.0 (main, Jan 10 2024, 14:25:36) [GCC 11.3.1 20221121 (Red Hat 11.3.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import multiprocessing
>>> multiprocessing.cpu_count()
32
>>> import psutil
>>> psutil.cpu_count(logical=False)
32
```

There must be some sort of hardware restriction that prevents process assignment to certain cores here.  In this cause we would want to use `oversubscribe` in the mpi command, however the `MPIBackend` won't apply that keyword because of the logic below. In my case my n.procs < 32 and so oversubscribe and hyper-threading will never be executed. 
https://github.com/jonescompneurolab/hnn-core/blob/ccef332d3190a6a7338703c74641abd817d3056d/hnn_core/parallel_backends.py#L609-L623

**Comment by rythorpe** [comment]
A few things:

1. I agree with @jasmainak that we should always default to `MPIBackend` for the GUI and use `JoblibBackend` as backup.
2. @dylansdaniels in the old GUI I always found setting the simulation hyperparameters distracting to the flow of our workshops, so yeah, I guess I'm advocating for limiting backend exposure as much as possible. Users who need more advanced control of backend parameters should be encouraged to use the HNN-core API directly.
3. @gtdang, I totally see your point that parallelizing with the maximal number of cores is not ideal. Given communication overhead, I've generally found that setting `n_procs` to >20 isn't beneficial for the default network. The other thing to consider that is in favor of exposing the backend API, is if we ever host a workshop on shared resources (e.g., as we have in the past with Metacell), it may at times be more efficient to have individual users run a smaller number of parallel processes so that the class, as a whole, doesn't get bogged down.

Aside from changing the default to `MPIBackend` for the GUI as mentioned above, what if we fix the logic within `MPIBackend` to always allow oversubscription and then provide a resonable default for `n_procs` in the GUI? If you still think it's necessary to expose the backend API in the GUI, I can get on-board, but I do think we should be mindful about limiting the number non-critical parameters in the GUI so that they're (at least visually) not overwhelming.

**Comment by dylansdaniels** [comment]
In additional Metacell, we also have the GUI available on other shared resources / HPCs like NSG and Oscar. In future workshops, I believe the plan is to encourage more people to start using the GUI on NSG, and I wonder if we'll run into the issue @gtdang found where MPI fails

**Comment by gtdang** [comment]
I don't think the default for the GUI should be a backend that is an optional dependency that you can't install with the package and which requires variable set-up depending on user hardware and operating system.

**Comment by rythorpe** [comment]
As @jasmainak suggested, the idea would be to default to `MPIBackend` only if it's available, and then switch to `JoblibBackend` if it's not.

Is there perhaps another way we can streamline use of the `MPIBackend` for the GUI? I don't think we've ever run a workshop without explicitly using this backend because simulations take too long without it. With the previous GUI, there were a handful of steps that added to the frustration (at least for me) of just getting a single simulation to run in real-time during workshops, one of which was setting the number of cores over which to parallelize the simulation. In service of the vast majority of GUI use-cases, it'd be nice to streamline this better. As I mentioned previously, adding the number of cores as a simulation parameter in the GUI isn't, in itself, a deal breaker, but I think there's something to be said for a GUI simulation that "just runs" quickly without having to sift through a myriad of parameters that don't directly relate to the scientific discourse of our workshops.

**Comment by gtdang** [comment]
OK. I think that check for mpi and changing the backend on the start-up should be easy to implement. 

I agree that a change to the `MPIBackend` class is needed, though I don't think we yet understand all the variability and system behaviors of different hardware set-ups to define default safe parameters and arguments with the `mpiexec` call. I only have been able to test my local macOS and Oscar's virtual machine instance GUI behavior. As the `MPIBackend` class is the core API, we ought to also look at how it behaves with different system architectures, operating systems. For example, just in the world of HPC, I'm not sure how it will behave in a virtual machine instance vs a Slurm job.  

Given that unknown I feel like it's the easiest for now to expose the GUI argument to allow users to try different values that work for their system environment.

**Comment by gtdang** [comment]
Added logic to check of packages for mpi use and setting the default backend. https://github.com/jonescompneurolab/hnn-core/pull/871/commits/b5a07da14b03bf69c51bb78b7cd7b60efc7cd299

File: hnn_core/gui/gui.py [hunk 2]
```diff
@@ -663,6 +694,10 @@ def compose(self, return_layout=True):
                 self.widget_ntrials, self.widget_backend_selection,
                 self._backend_config_out]),
         ], layout=self.layout['config_box'])
+        # Displays the default backend options
+        handle_backend_change(self.widget_backend_selection.value,
+                              self._backend_config_out, self.widget_mpi_cmd,
+                              self.widget_n_jobs)
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:700]: I was actually suggesting something simpler ... hnn-core already defaults to JoblibBackend if MPI is not available, so all you had to do was expose MPIBackend and get rid of the JoblibBackend. Anyhow @dylansdaniels do you want this merged for the GUI tutorial today?

Comment (gtdang) [reply hnn_core/gui/gui.py:700]: That handler in the gui changes the displayed options based on the backend drop-down it doesn't switch between backends. In hnn-core it MPIBackend only switches to joblib is it's a single process, I don't think it checks for MPI and reverts to joblib if it can't find it.

**Comment by rythorpe** [comment]
> OK. I think that check for mpi and changing the backend on the start-up should be easy to implement.
> 
> I agree that a change to the `MPIBackend` class is needed, though I don't think we yet understand all the variability and system behaviors of different hardware set-ups to define default safe parameters and arguments with the `mpiexec` call. I only have been able to test my local macOS and Oscar's virtual machine instance GUI behavior. As the `MPIBackend` class is the core API, we ought to also look at how it behaves with different system architectures, operating systems. For example, just in the world of HPC, I'm not sure how it will behave in a virtual machine instance vs a Slurm job.
> 
> Given that unknown I feel like it's the easiest for now to expose the GUI argument to allow users to try different values that work for their system environment.

That's fair. I still think it's non-ideal that MPI oversubscription isn't being invoked when it's supposed too, but we can fix this later if you think more investigation is needed.

File: hnn_core/gui/gui.py (hunk 2 above)
Comment (gtdang) [reply hnn_core/gui/gui.py:700]: Nevermind, you're right. It does revert to joblib. Though it runs on joblib with only 1 processor, is there a reason why it only uses 1?

Comment (rythorpe) [reply hnn_core/gui/gui.py:700]: I think it's because specifying multiple processes with the joblib backend, unlike the MPI backend, deosn't make sense for all types of simulations. Since joblib parallelizes over multiple trials, it only works for simulations that include multiple trials. Plus parallel joblib processes are memory intensive, so silently switching from MPI to joblib can produce errors if the user wasn't anticipating needing the extra resources needed for a joblib task with many trials.

Comment (jasmainak) [reply hnn_core/gui/gui.py:700]: I agree with @rythorpe . I would say the GUI should just use MPIBackend, and if not available, just run everything in a single core. It's probably a misnomer to call it "JoblibBackend" if it's running on a single core. Hence, I wouldn't bother exposing it to user.

File: hnn_core/gui/gui.py (hunk 1 above)
Comment (jasmainak) [reply hnn_core/gui/gui.py:326]: I thought you could adjust the number of cores in MPIBackend too. Is that not the case? I think n_cores - 1 may not be the best default ... but I'm not an MPI expert. I know it's a bad idea with joblibs

Comment (jasmainak) [reply hnn_core/gui/gui.py:326]: I would just set it to 1 rather than attempt to guess the total number of cores. During the tutorial, the user can be instructed to use more cores. The less magic there is in the code, the less likely it will fail.

File: hnn_core/gui/gui.py (hunk 2 above)
Comment (gtdang) [reply hnn_core/gui/gui.py:700]: Do we really want to strip the functionally of using multicores with joblib? It's a nice feature for those who don't have MPI implemented. On my machine a 3 trial simulation ran in less than half the time on 3 cores than 1 core.

Comment (jasmainak) [reply hnn_core/gui/gui.py:700]: The point is I have never seen anyone run multiple trials in the GUI ... but I will defer to the judgement of @rythorpe and @ntolley as they are the real users

Comment (rythorpe) [reply hnn_core/gui/gui.py:700]: I'm on the edge honestly. 9/10 times GUI users will be using `MPIBackend`, which applies to all types of simulations provided that it's installed correctly. IDK that providing granular control of `JoblibBackend` is worth it, but I could be convinced otherwise.

Comment (rythorpe) [reply hnn_core/gui/gui.py:700]: Do we know if anyone who is actually using `JoblibBackend` in the GUI for the their research? I personally don't know of anyone and am generally under the impression that folks who prefer the GUI would want a general-use parallel backend that works without too much fuss.

Your work @ntolley is unique in that your questions require lots of parameter sweeps, which are best served by using the API directly. Note that limited RAM is only an issue for Joblib, which I'm arguing should essentially default to running a single job within the context of the GUI.

Comment (ntolley) [reply hnn_core/gui/gui.py:700]: Pretty sure a lot of Darcy's initial work compared over multiple trials, although I'll concede she's using the API now. If you're doing 1-1 comparisons of ERP's for instance it's definitely more robust to compute an average over multiple trials. In tutorials we use 1 trial because it's efficient, but that isn't necessarily how you should fit these signals to data.

Comment (ntolley) [reply hnn_core/gui/gui.py:700]: Again I totally agree with making MPI with oversubscription the default, but I genuinely think trial-based parallelism is a useful feature even in the GUI setting! 

These are my cards on the table :smile:, but if I get outvoted I'll concede

Comment (rythorpe) [reply hnn_core/gui/gui.py:700]: Fair enough. If GUI users use it, we should support it.

Comment (jasmainak) [reply hnn_core/gui/gui.py:700]: I personally feel that once a feature is added, users find ways to justify keeping it. That's why one should start with minimal features first and then add based on necessity and feedback from user base. Anyway, call is yours @ntolley !

**Comment by asoplata** [comment]
I've made a PR to @gtdang 's branch here: https://github.com/brown-ccv/hnn-core-ccv/pull/2 which appears to work in and compensate for determining cores etc. in all of the expected use-cases . Pasting the list of use-cases here from the commit message:

In `_determine_cores_hwthreading()`, if you read the lengthy comments, I
have thought through each common scenario, and I believe resolved what
to do for each, with respect to the number of cores to use and whether
or not to use hardware-threading. These scenarios include: the user
choosing to use hardware-threading (default) or not, across Macos
variations with and without hardware-threading, Linux local computer
variations with and without hardware-threading, and Linux
HPC (e.g. OSCAR) variations which appear to never support
hardware-threading. In the Windows case, due to both https://github.com/jonescompneurolab/hnn-core/issues/589 and the
currently-untested MPI integration on Windows, I always report the
machine as not having hardware-threading.

Note that this does NOT change the default number of jobs to use for the
GUI if MPI is detected. Such a change breaks the current test_gui.py
testing: see https://github.com/jonescompneurolab/hnn-core/issues/960 .

**Comment by asoplata** [comment]
There is some git cleanup going to happen before this is merged: @gtdang is going to rebase his work onto a side branch on upstream, then I will apply my changes to that. This should make the result significantly easier to see the code changes.

**Comment by asoplata** [comment]
The code itself is now ready for merge (thanks for the typo fixing and docs George!), barring any unsaid comments by @ntolley et al. However, the deep dive I plan to do with @dylansdaniels and @ntolley still remains to be done. I have made a wiki page here https://github.com/jonescompneurolab/hnn-core/wiki/CPU-Core-Counting-results-(2025%E2%80%9001%E2%80%9017) where I have recorded output for every common usecase, including Logical, Physical, and Affinity core number (when applicable), along with what's returned with all variations of `_determine_cores_hwthreading()`. I believe this makes it clear that we get the core number we want in a configurable way, and in all common usecases (except for Windows, which is a special case where MPI isn't fully supported right now). Please review it.

File: hnn_core/parallel_backends.py [hunk 3]
```diff
@@ -554,34 +555,237 @@ def simulate(self, net, tstop, dt, n_trials, postproc=False):
...
+    ----------
+    enable_hwthreading : bool
+        Whether to detect support for hardware-threading and, if the feature is
+        detected, return the available number of 'logical' hardware-threaded
+        cores. Defaults to True. If 'False', or the feature is not detected,
+        return the available number of 'physical' cores (excluding
+        double-counting of hardware-threaded cores).
+    sensible_default_cores : bool
+        Whether to decrease the number of cores returned in a reasonable
+        manner, such that it balances speed with the user experience (e.g.,
```
Comment (ntolley) [inline hnn_core/parallel_backends.py:598]:
```suggestion
        manner (-1 from the detected number), such that it balances speed with the user experience (e.g.,
```

**Comment by asoplata** [comment]
Did a force-push (no conflicts/rewriting) just to make sure it passes our switch to `ruff check` from `flake8` and use current tests.

@ntolley @dylansdaniels At the risk of making things more complex and more verbose, I attempted to make things a little clearer in the most recent commit (here https://github.com/jonescompneurolab/hnn-core/pull/871/commits/eeff4de6e179dc102d4f86d46dbe468165799875 ) which is new. Namely, I changed the argument in `_determine_cores_hwthreading` from `enable_hwthreading` to `use_hwthreading_if_found` which I think is much more clear. 

I also tried to make `MPIBackend.__init__()`more explicit: instead of just having the arguments `hwthreading` and `oversubscribe` (which had the values of None, True, or False), it now has 4 arguments:
1. `use_hwthreading_if_found`: literally passed directly to `_determine_cores_hwthreading`
2. `sensible_default_cores`: literally passed directly to `_determine_cores_hwthreading`
3. `override_hwthreading_option`: this behaves like the old `hwthreading` argument, but is more clear. If None (default), the hwthreading option is determined by the `use_hwthreading_if_found` option. If True, then the user is "overriding" the hwthreading detection and the option is passed to MPI no matter what. If False, then similarly, the user is "overriding" the hwthreading detection and the option is never passed to MPI.
4. `override_oversubscribe_option`: same as `override_hwthreading_option` but for the oversubscribe option.

File: hnn_core/parallel_backends.py [hunk 4]
```diff
@@ -554,34 +555,269 @@ def simulate(self, net, tstop, dt, n_trials, postproc=False):
...
+        by default.
+    use_hwthreading_if_found : bool
+        Specifies whether the class should try to detect hardware-threading,
+        and, if it is found, then both use MPI's '--use-hwthread-cpus' option
+        and change the number of CPU cores used. Defaults to 'True'. Note that
+        this is passed to an option of the same name in
+        `_determine_cores_hwthreading`; see that function for more details.
+    sensible_default_cores : bool
+        Specifies whether to limit the number of CPU cores used based on a
+        "reasonable" heuristic. Defaults to 'False'. Note that this is passed
```
Comment (ntolley) [inline hnn_core/parallel_backends.py:800]: @asoplata can you specify exactly what is done? i.e. reduce core count by 1

File: hnn_core/parallel_backends.py [hunk 5]
```diff
@@ -594,59 +830,74 @@ class MPIBackend(object):
...
-            oversubscribe = True
-
-        hyperthreading = False
-
-        if _has_mpi4py() and _has_psutil():
-            import psutil
+        # Check of psutil and mpi4py import has been moved into this function,
+        # since this function is called by GUI before MPIBackend()
+        # instantiated.
+        [n_available_cores, hwthreading_available] = \
```
Comment (ntolley) [inline hnn_core/parallel_backends.py:852]: any reason for using brackets here? does ruff autoformat to this?

File: hnn_core/parallel_backends.py (hunk 4 above)
Comment (asoplata) [reply hnn_core/parallel_backends.py:800]: I originally wrote a longer description in the docstring for the `_determine...` function, which makes this inconsistent. I will add my expanded definition here as well.

File: hnn_core/parallel_backends.py (hunk 5 above)
Comment (asoplata) [reply hnn_core/parallel_backends.py:852]: I just do that out of habit, since it's easier to see that that the return is a list. I will remove it though.

File: hnn_core/parallel_backends.py (hunk 4 above)
Comment (asoplata) [reply hnn_core/parallel_backends.py]: I've hopefully resolved this in the latest commit

File: hnn_core/parallel_backends.py (hunk 5 above)
Comment (asoplata) [reply hnn_core/parallel_backends.py]: I've hopefully resolved this in the latest commit

File: hnn_core/tests/test_parallel_backends.py [hunk 6]
```diff
@@ -174,26 +198,93 @@ def test_run_mpibackend_oversubscribed(self, run_hnn_core_fixture):
...
+                    use_hwthreading_if_found=use_hwthreading_if_found
+            ) as backend:
+                assert backend.n_procs == oversubscribed_procs
+                assert "--oversubscribe" in ' '.join(backend.mpi_cmd)
+                if detected_hwthreading:
+                    assert "--use-hwthread-cpus" in ' '.join(backend.mpi_cmd)
+                simulate_dipole(net, tstop=40)
+
+        # Check that the simulation fails if oversubscribe is forced off
+        with pytest.warns(UserWarning) as record:
```
[the remark this answers is not in the document]
Comment (asoplata) [reply hnn_core/tests/test_parallel_backends.py]: I've removed this in the latest commit. I suspect that I copied over the "warning wrapper" from a pre-existing test, then changed the test inside, but forgot to remove the `with Warning` that become unnecessary.

File: hnn_core/parallel_backends.py [hunk 7]
```diff
@@ -594,59 +836,74 @@ class MPIBackend(object):
...
+            (
+                (override_oversubscribe_option is None) and
+                (self.n_procs > n_available_cores)
+        )):
+            warn(
+                "Number of requested MPI processes exceeds available "
+                "cores. Enabling MPI oversubscription automatically."
+            )
+            self.mpi_cmd += " --oversubscribe"
+
```
Comment (dylansdaniels) [inline hnn_core/parallel_backends.py:896]: Possibly add a warning in the edge case where the user specifies override_oversubscribe = False and manually sets n_procs to a value > n_available_cores

File: hnn_core/tests/test_parallel_backends.py [hunk 8]
```diff
@@ -162,7 +184,9 @@ def test_terminate_mpibackend(self, run_hnn_core_fixture):
 
     @requires_mpi4py
     @requires_psutil
-    def test_run_mpibackend_oversubscribed(self, run_hnn_core_fixture):
+    @pytest.mark.parametrize("use_hwthreading_if_found", [True, False])
+    def test_run_mpibackend_oversubscribed(self, run_hnn_core_fixture,
```
[the remark this answers is not in the document]
Comment (asoplata) [reply hnn_core/tests/test_parallel_backends.py]: I addressed this in the latest commit: Now, the tests are separated into those relevant to oversubscription in one function, and those relevant to hwthreading in another function. Also, the current option/case being tested is significantly more clear.

File: hnn_core/parallel_backends.py (hunk 7 above)
Comment (asoplata) [reply hnn_core/parallel_backends.py:910]: Added in latest commit.

**Comment by asoplata** [comment]
@dylansdaniels Can you please merge this if you approve of the changes I made in response to your comments? I ask because there is another new PR #998 which wants to make changes to much of the nearby code.

File: hnn_core/tests/test_parallel_backends.py [hunk 9]
```diff
@@ -240,16 +240,18 @@ def test_run_mpibackend_oversubscribed(self, use_hwthreading_if_found):
                 assert "--oversubscribe" in ' '.join(backend.mpi_cmd)
                 simulate_dipole(net, tstop=40)
 
-        # Case 2: Check that the simulation runs, but with no warning if
-        # oversubscription is forced on
+        # Case 2: Check that no warning, but option passed if oversubscription
```
[the remark this answers is not in the document]
Comment (asoplata) [reply hnn_core/tests/test_parallel_backends.py]: Done and pushed

**Comment by asoplata** [comment]
Needed to force-push again in order to rebase onto master to compensate for #994. This latest version should work fine. Sidenote that, for all PRs which existed before #994, if there is new code pushed, they will ALL have to be rebased to a post-#994 commit in order for their tests to pass.
